routes/user.js

```diff
@@ -2,100 +2,123 @@
-        // eslint-disable-next-line no-process-env
-        link: process.env.DEBUG_DUMMY_SMTP ? url : undefined,
-      });
-    })
-    .catch(error => {
-      res.json({ email, info: "failed to send verification link", error });
-      throw error;
+
+  if (!email)  {
+    resDatabaseError(res);
+    return;
+  }
+
+  const url = `${AppConstants.SERVER_URL}/user/verify?state=${state}&email=${email}`;
+
+  const { error } = await Subscribers.addTempUser(email, state);
+
+  if (error) {
+    resDatabaseError(res);
+    return;
+  }
+
+  try {
+    await EmailUtils.sendEmail(email, "Firefox Breach Alert",
+      `Visit this link to subscribe: ${url}`);
+
+    res.status(202).json({
+      code: 202,
+      info: "Sent verification link",
+      // Send the would-be link back to the client in dummy mode.
+      // eslint-disable-next-line no-process-env
+      link: process.env.DEBUG_DUMMY_SMTP ? url : undefined,
+    });
+  } catch (e) {
+    console.log(e);
+    res.status(500).json({
+      code: 500,
+      error_code: ResponseCodes.InternalError,
+      info: "SMTP error."
     });
+  }
 });
 
-router.get("/verify", (req, res) => {
-  const email = gUnverifiedEmails.get(req.query.state);
-  if (email) {
-    gUnverifiedEmails.delete(req.query.state);
-    Subscribers.addUser(email).then(() => {
-      res.json({ email, info: `Successfully added ${email}`});
-    }).catch((error) => {
-      res.json({ info: "Failed to add email to database.", error });
+router.get("/verify", async (req, res) => {
+  let email, token, error, duplicate;
+  ({ email, token, error } = await Subscribers.getTempUser(req.query.email));
+
+  if (error) {
+    resDatabaseError(res);
+    return;
+  }
+
+  if (!email) {
+    res.status(400).json({
+      code: 400,
+      error_code: ResponseCodes.EmailNotFound,
+      info: "Email not marked for verification."
     });
     return;
   }
-  // TODO: Needs better error message.
-  res.json({ info: "Who are you?" });
+
+  // Invalidate the entry even if the token doesn't match.
+  ({ error } = await Subscribers.deleteTempUser(email));
+  if (error) {
+    resDatabaseError(res);
+    return;
+  }
+
+  if (token !== req.query.state) {
+    res.status(400).json({
+      code: 400,
+      error_code: ResponseCodes.TokenMismatch,
+      info: "Email and token do not match."
+    });
+    return;
+  }
+
+  ({ error, duplicate } = await Subscribers.addUser(email));
+  if (error) {
+    resDatabaseError(res);
+    return;
+  }
+
+  res.status(201).json({
+    code: 201,
+    duplicate,
+    info: `Successfully added ${email}`
+  });
 });
 
-router.post("/remove", (req, res) => {
+router.post("/remove", async (req, res) => {
   const email = req.body.email;
-  Subscribers.deleteUser(email).then(() => {
+  try {
+    await Subscribers.deleteUser(email);
     res.json({ email, info: "removed user" });
-  }).catch((error) => {
+  } catch (error) {
     res.json({ email, info: "Failed to remove user.", error });
```

Comment:
Should this set a custom `.status()`?

Reply:
I'm going to fix up the /remove API in a future commit. It needs email verification of some sort too.